plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,511 @@
+                    saveLatestDeployRetryTimesStep::onFailure
+                );
+        }, getDeployRetryTimesStep::onFailure);
+    }
+
+    /**
+     * auto undeploy all the models under the node id <br>
+     *
+     * @param localNodeId local node id
+     * @param modelIdAndNodeIds model ids
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void unDeployModel(String localNodeId, Map<String, Set<String>> modelIdAndNodeIds) throws MLException {
+        StepListener<MLUndeployModelNodesResponse> unDeployModelStep = new StepListener<>();
+
+        ActionListener<MLUndeployModelNodesResponse> listener = ActionListener
+            .wrap(
+                response -> log.info("the models is auto undeploying under the node {} ", localNodeId),
+                exception -> log.error("fail to undeploy models under the node id: {} \nthe reason is: {}", localNodeId, exception)
+            );
+        String[] modelIds = modelIdAndNodeIds.keySet().toArray(new String[] {});
+
+        autoUnDeployModelByNodeAndModelId(localNodeId, modelIds, listener);
+
+        unDeployModelStep.whenComplete(unDeployModelStep::onResponse, unDeployModelStep::onFailure);
+    }
+
+    /**
+     * auto deploy 1 model under the local node id <br>
+     *
+     * @param nodeIds node ids
+     * @param modelId model id
+     * @param listener listener
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoDeployModelByNodeAndModelId(String[] nodeIds, String modelId, ActionListener<MLDeployModelResponse> listener)
+        throws MLException {
+        MLDeployModelRequest deployModelRequest = new MLDeployModelRequest(modelId, nodeIds, false, false);
+
+        client.execute(MLDeployModelAction.INSTANCE, deployModelRequest, listener);
+    }
+
+    /**
+     * auto undeploy 1 model under the local node id <br>
+     *
+     * @param localNodeId node id
+     * @param modelIds all the model id
+     * @param listener listener
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoUnDeployModelByNodeAndModelId(String localNodeId, String[] modelIds, ActionListener<MLUndeployModelNodesResponse> listener)
+        throws MLException {
+        MLUndeployModelNodesRequest undeployModelNodesRequest = new MLUndeployModelNodesRequest(new String[] { localNodeId }, modelIds);
+
+        client.execute(MLUndeployModelAction.INSTANCE, undeployModelNodesRequest, listener);
+    }
+
+    /**
+     * get all models from the index <br>
+     *
+     * @param searchResponseActionListener listener for search response
+     */
+    @VisibleForTesting
+    void queryModels(ActionListener<SearchResponse> searchResponseActionListener) {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
+        QueryBuilder queryBuilder = QueryBuilders.boolQuery().mustNot(QueryBuilders.existsQuery(MLModel.CHUNK_NUMBER_FIELD));
```

Comment:
Seems no filter for model state. So we will query all models?

Reply:
Yes, our Shanghai team is discussing that all models need to be re-deployed

Reply:
I will changed it,let the model which the state is REGISTERED,UPLOADED,DEPLOY_FAILED,LOAD_FAILED,UNDEPLOYED,UNLOADED will not be queried~